src/library/crate/cratefeature.cpp

```diff
@@ -0,0 +1,915 @@
+                return;
+            }
+            if (!crate.hasName()) {
+                QMessageBox::warning(
+                        nullptr,
+                        tr("Duplicating Crate Failed"),
+                        tr("A crate cannot have a blank name."));
+                continue;
+            }
+            if (m_pTrackCollection->crates().readCrateByName(crate.getName())) {
+                QMessageBox::warning(
+                        nullptr,
+                        tr("Duplicating Crate Failed"),
+                        tr("A crate by that name already exists."));
+                crate.resetName();
+                continue;
+            }
+        }
+
+        CrateId crateId;
+        if (m_pTrackCollection->insertCrate(crate, &crateId)) {
+            QList<TrackId> trackIds;
+            trackIds.reserve(
+                    m_pTrackCollection->crates().countCrateTracks(oldCrate.getId()));
+            {
+                CrateTrackSelectResult crateTracks(
+                        m_pTrackCollection->crates().selectCrateTracksSorted(oldCrate.getId()));
+                while (crateTracks.next()) {
+                    trackIds.append(crateTracks.trackId());
+                }
+            }
+            m_pTrackCollection->addCrateTracks(crateId, trackIds);
+            activateCrate(crateId);
+        } else {
+            qDebug() << "Error creating crate with name " << crate.getName();
+            QMessageBox::warning(
+                    nullptr,
+                    tr("Creating Crate Failed"),
+                    tr("An unknown error occurred while creating crate: ") + crate.getName());
+        }
+    } else {
+        qDebug() << "Failed to duplicate selected crate";
+    }
+}
+
+void CrateFeature::slotToggleCrateLock() {
+    Crate crate;
+    if (readLastRightClickedCrate(&crate)) {
+        crate.setLocked(!crate.isLocked());
+        if (!m_pTrackCollection->updateCrate(crate)) {
+            qDebug() << "Failed to toggle lock of crate" << crate;
+        }
+    } else {
+        qDebug() << "Failed to toggle lock of selected crate";
+    }
+}
+
+void CrateFeature::slotAutoDjTrackSourceChanged() {
+    Crate crate;
+    if (readLastRightClickedCrate(&crate)) {
+        if (crate.isAutoDjSource() != m_pAutoDjTrackSourceAction->isChecked()) {
+            crate.setAutoDjSource(m_pAutoDjTrackSourceAction->isChecked());
+            m_pTrackCollection->updateCrate(crate);
+        }
+    }
+}
+
+QModelIndex CrateFeature::rebuildChildModel(CrateId selectedCrateId) {
+    TreeItem* pRootItem = m_childModel.getRootItem();
+    VERIFY_OR_DEBUG_ASSERT(pRootItem != nullptr) {
+        return QModelIndex();
+    }
+    m_childModel.removeRows(0, pRootItem->childRows());
+
+    TrackId selectedTrackId =
+            m_pSelectedTrack ? m_pSelectedTrack->getId() : TrackId();
+
+    QList<TreeItem*> modelRows;
+    modelRows.reserve(m_pTrackCollection->crates().countCrates());
```

Comment:
I am nor sure if it is worth the work, but we can get rid of this intermediated List, if we merge  m_childModel.insertRows into this function or if we move this function to CrateSummarySelectResult or a kind of wrapper on it that is passed to 
m_childModel.insertRows.